### DISCUSSION TITLE
How to compress WASM app?

### DISCUSSION THREAD

**Discussion by dbriard** [AvaloniaUI/Avalonia#11499] [Q&A] [answered]
Hi, I know that WASM applications are generally compressed (gzip or brotli) to speedup loading on slow connection computers, but I cannot find any ressources how to do that? I read some articles about Blazor and it seems the compression is done at the publish stage, but I don't know how that apply to Avalonia. 

Anyone know steps, samples, or where I can find a good tutorial for compressing an Avalonia v11 WASM project? My AppBundle is around 80MB and I can zip it to 24 MB. Can I get something like that after compression?

Thanks!

**Comment by wieslawsoltes** [comment]
Checkout this script: https://github.com/wieslawsoltes/ChatGPT/blob/d268610da16050187705f584b8c5ac6a62e69b9a/.github/workflows/pages.yml#L35

It uses: https://github.com/wieslawsoltes/DotNetCompress

**Comment by dbriard** [comment]
Hi @wieslawsoltes 

Thanks a lot for the sample.
The script is to automatically deploy on github pages? I am not familiar with that but that is very interesting, I'll try to add that to my project.

So at the moment, I can compress thanks to your DotNetCompress and I get shrinked .gz and .bt files along with the regular files.
I do not see any new lines of code in index.html or main.js in your project, and just adding the .bt and .gz files in the pages do not seems to work.

Could you give a little more details please? May be something to configure in github?
If I understand correctly, it seems that Github compress on the fly dotnet.wasm with gzip, but it is the only file that is compressed when I look at the Inspector. And it do not use the precompressed .gz file.

**Comment by dbriard** [comment]
@wieslawsoltes Okay, I noticed that your ChatGPT page also do not serve .br or .gz files. For instance, the Avalonia.Fonts.Inter.dll is downloaded as 2.5MB (uncompressed) and there is nothing in the response header about compression.

So it seems that Github do not supports pre-compressed files, and only dotnet.wasm is gzip compressed on the fly by githut.

I'll continue to investigate and try to add script that download and decompress .br files on client side.

**Comment by danwalmsley** [comment, marked as answer]
@dbriard for the .br files to be served depends on the host server. Github pages doesnt support it.

Iv used azure static web apps for https://solitaire.xaml.live, if the file.extension.br exists when file.extension is requested by the browser, azure serves the .br version.

This is basically essential for wasm apps.

Check if your host can enable brotli compression, it may be an configuration setting to turn it on.

**Reply by dbriard** [reply]
Thanks for the info, at the moment I was just testing on GitHub but I may try azure if that support .be.

..and beautiful App of your! :)

**Comment by IvanJosipovic** [comment]
On the Blazor WASM side, there is way to force the JS to load Brotli or GZIP resources, see here https://learn.microsoft.com/en-us/aspnet/core/blazor/host-and-deploy/webassembly?view=aspnetcore-8.0#compression

Is there a similar approach on the Avalonia side?

**Reply by maxkatz6** [reply]
1. Rely on static server compression, which normally works pretty well. But not everywhere is available.
2. Use [DotNetCompress](https://github.com/wieslawsoltes/DotNetCompress) or other manual compression tools compile time.
3. With .NET 8 Microsoft had backported compression from Blazor to normal WASM apps. [This PR](https://github.com/AvaloniaUI/Avalonia/pull/14225) is required to make it compatible with Avalonia.
